File: agent/model_adapter.py
from __future__ import annotations

import os
from typing import Any, Protocol

import httpx

from schemas.agent import ChatRequest
from tools.registry import ToolRegistry


class ModelClient(Protocol):
    async def complete(self, request: ChatRequest, registry: ToolRegistry) -> str:
        """Return a model answer for the request."""


class MockModelClient:
    async def complete(self, request: ChatRequest, registry: ToolRegistry) -> str:
        selected = registry.select_for_message(request.message) or registry.names()[:5]
        summaries = []
        for tool_name in selected:
            result = await registry.run(tool_name)
            summaries.append(f"{tool_name}: {result['summary']}")

        return (
            "Mock MiniMax AIOps response. "
            "I correlated read-only AWS platform signals across the simulated accounts. "
            + " ".join(summaries)
        )


class MiniMaxChatCompletionsAdapter:
    """MiniMax OpenAI-compatible Chat Completions adapter."""

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.minimax.io",
        model: str = "MiniMax-M3",
        timeout_seconds: float = 30.0,
    ) -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.timeout_seconds = timeout_seconds

    async def complete(self, request: ChatRequest, registry: ToolRegistry) -> str:
        tool_context = []
        for tool_name in registry.select_for_message(request.message) or registry.names()[:5]:
            result = await registry.run(tool_name)
            tool_context.append(
                {
                    "tool": tool_name,
                    "summary": result["summary"],
                    "data": result["data"],
                }
            )

        payload: dict[str, Any] = {
            "model": self.model,
            "messages": [
                {
                    "role": "system",
                    "content": (
                        "You are a read-only AWS AIOps Platform Agent. "
                        "Use supplied tool context to diagnose, summarize impact, "
                        "recommend safe next steps, and never claim to mutate AWS resources."
                    ),
                },
                {
                    "role": "user",
                    "content": (
                        f"User request: {request.message}\n\n"
                        f"Read-only tool context: {tool_context}"
                    ),
                },
            ],
            "thinking": {"type": "disabled"},
            "temperature": 0.2,
        }

        async with httpx.AsyncClient(timeout=self.timeout_seconds) as client:
            response = await client.post(
                f"{self.base_url}/v1/chat/completions",
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                },
                json=payload,
            )
            response.raise_for_status()
            body = response.json()

        return body["choices"][0]["message"]["content"]


def get_model_client() -> ModelClient:
    api_key = os.getenv("MINIMAX_API_KEY")
    if api_key:
        return MiniMaxChatCompletionsAdapter(
            api_key=api_key,
            base_url=os.getenv("MINIMAX_BASE_URL", "https://api.minimax.io"),
            model=os.getenv("MINIMAX_MODEL", "MiniMax-M3"),
        )
    return MockModelClient()

File: src/agent/modelAdapter.ts
import type { ChatRequest } from "../schemas/agent.js";
import type { ToolRegistry } from "../tools/registry.js";

export interface ModelClient {
  complete(request: ChatRequest, registry: ToolRegistry): Promise<string>;
}

export class MockModelClient implements ModelClient {
  async complete(request: ChatRequest, registry: ToolRegistry): Promise<string> {
    const selected = registry.selectForMessage(request.message);
    const toolNames = selected.length > 0 ? selected : registry.names().slice(0, 5);
    const summaries: string[] = [];

    for (const toolName of toolNames) {
      const result = await registry.run(toolName);
      summaries.push(`${toolName}: ${result.summary}`);
    }

    return [
      "Mock local OpenAI-compatible AIOps response.",
      "I correlated read-only AWS platform signals across the simulated accounts.",
      ...summaries,
    ].join(" ");
  }
}

export class OpenAICompatibleChatAdapter implements ModelClient {
  constructor(
    private readonly baseUrl: string,
    private readonly model: string,
    private readonly apiKey?: string,
    private readonly providerName = "local OpenAI-compatible model",
  ) {}

  async complete(request: ChatRequest, registry: ToolRegistry): Promise<string> {
    const selected = registry.selectForMessage(request.message);
    const toolNames = selected.length > 0 ? selected : registry.names().slice(0, 5);
    const toolContext = [];

    for (const toolName of toolNames) {
      const result = await registry.run(toolName);
      toolContext.push({
        tool: toolName,
        summary: result.summary,
        data: result.data,
      });
    }

    const response = await fetch(`${this.baseUrl.replace(/\/$/, "")}/v1/chat/completions`, {
      method: "POST",
      headers: {
        ...(this.apiKey ? { Authorization: `Bearer ${this.apiKey}` } : {}),
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: this.model,
        messages: [
          {
            role: "system",
            content:
              "You are a read-only AWS AIOps Platform Agent. Use supplied tool context to diagnose, summarize impact, recommend safe next steps, and never claim to mutate AWS resources.",
          },
          {
            role: "user",
            content: `User request: ${request.message}\n\nRead-only tool context: ${JSON.stringify(toolContext)}`,
          },
        ],
        thinking: { type: "disabled" },
        temperature: 0.2,
      }),
    });

    if (!response.ok) {
      throw new Error(`${this.providerName} request failed with status ${response.status}.`);
    }

    const body = (await response.json()) as {
      choices?: Array<{ message?: { content?: string } }>;
    };
    const content = body.choices?.[0]?.message?.content;
    if (!content) {
      throw new Error(`${this.providerName} response did not include message content.`);
    }
    return content;
  }
}

export class MiniMaxChatCompletionsAdapter extends OpenAICompatibleChatAdapter {
  constructor(
    apiKey: string,
    baseUrl = "https://api.minimax.io",
    model = "MiniMax-M3",
  ) {
    super(baseUrl, model, apiKey, "MiniMax");
  }
}

export class LocalOpenAICompatibleAdapter extends OpenAICompatibleChatAdapter {
  constructor(
    baseUrl = "http://127.0.0.1:11434",
    model = "qwen2.5",
    apiKey?: string,
  ) {
    super(baseUrl, model, apiKey, "local OpenAI-compatible model");
  }
}

export function getModelClient(): ModelClient {
  const provider = process.env.MODEL_PROVIDER ?? "local";
  if (provider === "local") {
    return new LocalOpenAICompatibleAdapter(
      process.env.LOCAL_MODEL_BASE_URL ?? "http://127.0.0.1:11434",
      process.env.LOCAL_MODEL_NAME ?? "qwen2.5",
      process.env.LOCAL_MODEL_API_KEY,
    );
  }

  if (provider === "minimax" && process.env.MINIMAX_API_KEY) {
    return new MiniMaxChatCompletionsAdapter(
      process.env.MINIMAX_API_KEY,
      process.env.MINIMAX_BASE_URL ?? "https://api.minimax.io",
      process.env.MINIMAX_MODEL ?? "MiniMax-M3",
    );
  }

  return new MockModelClient();
}
